### PR TITLE
[wip] better integration with Delft via JEP

### PR DESCRIPTION
**Pull request by lfoppiano**
- support for anaconda and virtualenv 
 - support for mac and Linux (for Windows we need help)
 - automatic selection of python version and jep library (system/env installed or from grobid-home)
 - overall simplification of the installation process 
 - documentation

### CODE REVIEW

File: grobid-home/config/grobid.properties [hunk 1]
```diff
@@ -31,11 +31,12 @@ grobid.proxy_port=null
 #------------------------------------------------------
 
 #-------------------- runtime ------------------
-grobid.crf.engine=wapiti
-#grobid.crf.engine=delft
+#grobid.crf.engine=wapiti
+grobid.crf.engine=delft
```
Comment (de-code) [inline grobid-home/config/grobid.properties]: I guess that should be reverted before merging

**Comment by de-code** [comment]
have you looked into why the tests are failing?

**Comment by lfoppiano** [comment]
@de-code not yet, the problem with the tests is that some are specific for wapiti.. so at the moment is complicated to fix them. Lets make it work first with delft (and submodules) and then we can think of the tests.

**Comment by lfoppiano** [comment]
@de-code this morning I did some more tests, the outcome is confusing... 

If I install jep with pip in my virtual environment, at the moment of loading the library the whole JVM crashes (with both Oracle or OpenJDK jvm): 

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGBUS (0xa) at pc=0x0000000130e815d0, pid=82276, tid=47875
#
# JRE version: OpenJDK Runtime Environment (11.0.2+9) (build 11.0.2+9)
# Java VM: OpenJDK 64-Bit Server VM (11.0.2+9, mixed mode, tiered, compressed oops, g1 gc, bsd-amd64)
# Problematic frame:
# C  [libopenblas.dylib+0x1515d0]  dgemm_thread_nn+0x630
#
# No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /Users/lfoppiano/development/projects/delft/hs_err_pid82276.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
```
If I uninstall jep from the virtual environment it loads the library without errors, but it's not picking up the jep library from the native libraries in `grobid-home`. 


Finally, a good news, If the virtual environment is not specified in the command line or in the ENV var, everything fails as it should be.

**Comment by lfoppiano** [comment]
@de-code `jepPath` in `PythonEnvironmentConfig` should be checked that exists, I think. I see as the virtualenv is found, it is assumed that jep is installed. Shouldn't we load the shipped library if not present in the virtual env?

**Comment by de-code** [comment]
> If I install jep with pip in my virtual environment, at the moment of loading the library the whole JVM crashes (with both Oracle or OpenJDK jvm):

Yes, I observed the same. Apologies, if I hadn't made it clear. Which led me to suggest to not try to support a virtual env config in grobid.properties. One of the main problems is I think, that we don't start a new process and therefore we can't fool the process completely. e.g. (and that is really just one of the examples) we are hacking our way into convincing Java to load libraries from another path - but that doesn't include libraries loaded by the library itself (I think). It doesn't seem to be worth the hassle and the fairly unstable integration to me. It is working fine, if the virtual env was activated before.

> @de-code `jepPath` in `PythonEnvironmentConfig` should be checked that exists, I think. I see as the virtualenv is found, it is assumed that jep is installed. Shouldn't we load the shipped library if not present in the virtual env?

Yes, I was thinking whether that should be added. It would be good if it checked it. But ultimately it doesn't really change anything. If the directory doesn't exist or doesn't contain a library it won't load a library from that directory, same as if we didn't include. But it seems good practice to not add the directory.

**Comment by lfoppiano** [comment]
@de-code I've tried activating the conda environment and then running the service and I got the same issue. The service started without problems (the loading seems fine) but then it crashed the JVM at the first attempt to process something. 
Out of any other options I follow plan B (Plan A is rebooting my machine :-) ), so I checked out the error and installed libopenblas and magically the problem was solved... 😮 

I've installed version `0.3.6`. 

If I deactivate the conda env and try to run it it doesn't work. It used to work but I don't remember how I did it, so I think we can say that the env needs to be activated (as it's right to be, for any python package)

**Comment by lfoppiano** [comment]
OK I've got excited too easily... I've done some updates with anaconda and now I get a nice surprise: 

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00000001329115d3, pid=1251, tid=82435
#
# JRE version: OpenJDK Runtime Environment (11.0.2+9) (build 11.0.2+9)
# Java VM: OpenJDK 64-Bit Server VM (11.0.2+9, mixed mode, tiered, compressed oops, g1 gc, bsd-amd64)
# Problematic frame:
# C  [_multiarray_umath.cpython-36m-darwin.so+0x1205d3]  PyArray_Item_INCREF+0x43
#
# No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /Users/lfoppiano/development/projects/grobid/hs_err_pid1251.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
```

**Comment by lfoppiano** [comment]
OK, again me... spamming. 

In the last desperate attempt, I've updated conda-forge ... and now magically it doesn't crash anymore... 

not a clue what was the problem, well this is python so anything can be breaking anytime... 

```
(tensorflow) lucas-mbp:grobid lfoppiano$ conda list
# packages in environment at /anaconda3/envs/tensorflow:
#
# Name                    Version                   Build  Channel
absl-py                   0.7.1                    py36_0    conda-forge
appdirs                   1.4.3                      py_1    conda-forge
appnope                   0.1.0                 py36_1000    conda-forge
asn1crypto                0.24.0                py36_1003    conda-forge
astor                     0.7.1                      py_0    conda-forge
attrs                     19.1.0                     py_0    conda-forge
autopep8                  1.4.4                      py_0    conda-forge
backcall                  0.1.0                      py_0    conda-forge
beautifulsoup4            4.7.1                 py36_1001    conda-forge
black                     19.3b0                     py_0    conda-forge
blas                      2.10                   openblas    conda-forge
bleach                    3.1.0                      py_0    conda-forge
blinker                   1.4                        py_1    conda-forge
boto                      2.49.0                     py_0    conda-forge
boto3                     1.9.184                    py_0    conda-forge
botocore                  1.12.184                   py_0    conda-forge
bz2file                   0.98                       py_0    conda-forge
bzip2                     1.0.6             h1de35cc_1002    conda-forge
c-ares                    1.15.0            h1de35cc_1001    conda-forge
ca-certificates           2019.6.16            hecc5488_0    conda-forge
cctools                   895                  h7512d6f_0    conda-forge
certifi                   2019.6.16                py36_0    conda-forge
cffi                      1.12.3           py36hccf1714_0    conda-forge
chardet                   3.0.4                 py36_1003    conda-forge
chemdataextractor         1.3.0                      py_0    conda-forge
clang                     4.0.1                h662ec87_0  
clang_osx-64              4.0.1               h1ce6c1d_16  
clangxx                   4.0.1                hc9b4283_0  
clangxx_osx-64            4.0.1               h22b1bf0_16  
click                     7.0                        py_0    conda-forge
compiler-rt               4.0.1                h5487866_0  
conda                     4.7.5                    py36_0    conda-forge
conda-build               3.17.8                   py36_1    conda-forge
conda-package-handling    1.3.10                   py36_0    conda-forge
cryptography              2.7              py36h212c5bf_0    conda-forge
cssselect                 1.0.3                      py_0    conda-forge
cycler                    0.10.0                     py_1    conda-forge
dawg                      0.7.8            py36hfc679d8_0    conda-forge
dbus                      1.13.6               h2f22bb5_0    conda-forge
decorator                 4.4.0                      py_0    conda-forge
defusedxml                0.5.0                      py_1    conda-forge
docutils                  0.14                  py36_1001    conda-forge
entrypoints               0.3                   py36_1000    conda-forge
expat                     2.2.5             h6de7cb9_1003    conda-forge
fastcache                 1.1.0            py36h01d97ff_0    conda-forge
filelock                  3.0.10                     py_0    conda-forge
freetype                  2.10.0               h24853df_0    conda-forge
future                    0.17.1                py36_1000    conda-forge
gast                      0.2.2                      py_0    conda-forge
gensim                    3.4.0            py36h1de35cc_0  
gettext                   0.19.8.1          h46ab8bc_1002    conda-forge
glib                      2.58.3            h9d45998_1001    conda-forge
glob2                     0.7                        py_0    conda-forge
gmp                       6.1.2             h0a44026_1000    conda-forge
gmpy2                     2.1.0b1          py36h1744cca_0    conda-forge
grpcio                    1.16.1           py36h044775b_1  
h5py                      2.9.0           nompi_py36h35dc34b_1102    conda-forge
hdf5                      1.10.4          nompi_h0cbb7df_1106    conda-forge
hyperopt                  0.1.2                      py_0    conda-forge
icu                       58.2              h0a44026_1000    conda-forge
idna                      2.8                   py36_1000    conda-forge
imageio                   2.5.0                    py36_0    conda-forge
ipykernel                 5.1.1            py36h24bf2e0_0    conda-forge
ipython                   7.6.1            py36h5ca1d4c_0    conda-forge
ipython_genutils          0.2.0                      py_1    conda-forge
ipywidgets                7.5.0                      py_0    conda-forge
jedi                      0.14.0                   py36_0    conda-forge
jep                       3.8.2                    pypi_0    pypi
jinja2                    2.10.1                     py_0    conda-forge
jmespath                  0.9.4                      py_0    conda-forge
joblib                    0.13.2                     py_0    conda-forge
jpeg                      9c                h1de35cc_1001    conda-forge
json5                     0.8.5                      py_0    conda-forge
jsonschema                3.0.1                    py36_0    conda-forge
jupyter                   1.0.0                      py_2    conda-forge
jupyter_client            5.3.0                      py_0    conda-forge
jupyter_console           6.0.0                      py_0    conda-forge
jupyter_core              4.4.0                      py_0    conda-forge
jupyterlab                1.0.1                    py36_0    conda-forge
jupyterlab_server         1.0.0                      py_1    conda-forge
keras                     2.2.4                    py36_1    conda-forge
keras-applications        1.0.7                      py_1    conda-forge
keras-bert                0.53.0                   pypi_0    pypi
keras-embed-sim           0.4.0                    pypi_0    pypi
keras-layer-normalization 0.12.0                   pypi_0    pypi
keras-multi-head          0.20.0                   pypi_0    pypi
keras-pos-embd            0.10.0                   pypi_0    pypi
keras-position-wise-feed-forward 0.5.0                    pypi_0    pypi
keras-preprocessing       1.0.9                      py_1    conda-forge
keras-self-attention      0.41.0                   pypi_0    pypi
keras-transformer         0.23.0                   pypi_0    pypi
kiwisolver                1.1.0            py36h770b8ee_0    conda-forge
langdetect                1.0.7                      py_0    conda-forge
ld64                      274.2                h7c2db76_0    conda-forge
libarchive                3.3.3             h5c473cc_1006    conda-forge
libblas                   3.8.0               10_openblas    conda-forge
libcblas                  3.8.0               10_openblas    conda-forge
libcxx                    4.0.1                h579ed51_0  
libcxxabi                 4.0.1                hebd6815_0  
libffi                    3.2.1             h6de7cb9_1006    conda-forge
libgfortran               3.0.1                         0    conda-forge
libgpuarray               0.7.6             h1de35cc_1003    conda-forge
libiconv                  1.15              h01d97ff_1005    conda-forge
liblapack                 3.8.0               10_openblas    conda-forge
liblapacke                3.8.0               10_openblas    conda-forge
liblief                   0.9.0                h2a1bed3_1    conda-forge
libopenblas               0.3.6                hd44dcd8_4    conda-forge
libpng                    1.6.37               h2573ce8_0    conda-forge
libprotobuf               3.8.0                hfbae3c0_0    conda-forge
libsodium                 1.0.16            h1de35cc_1001    conda-forge
libtiff                   4.0.10            hd08fb8f_1003    conda-forge
libxml2                   2.9.9                hd80cff7_1    conda-forge
libxslt                   1.1.32            h1e79283_1003    conda-forge
llvm                      4.0.1                hc748206_0    conda-forge
llvm-lto-tapi             4.0.1                h6701bc3_0    conda-forge
llvm-openmp               8.0.0                h770b8ee_0    conda-forge
lxml                      4.3.4            py36h08abf6f_0    conda-forge
lz4-c                     1.8.3             h6de7cb9_1001    conda-forge
lzo                       2.10              h1de35cc_1000    conda-forge
mako                      1.0.10                     py_0    conda-forge
markdown                  3.1.1                      py_0    conda-forge
markupsafe                1.1.1            py36h1de35cc_0    conda-forge
mat2vec                   0.2                      pypi_0    pypi
matplotlib                3.1.1                    py36_0    conda-forge
matplotlib-base           3.1.1            py36h3a684a6_0    conda-forge
mistune                   0.8.4           py36h1de35cc_1000    conda-forge
monty                     1.0.6                    py36_1    matsci
mpc                       1.1.0             h3aece09_1006    conda-forge
mpfr                      4.0.2                h0a95893_0    conda-forge
mpmath                    1.1.0                      py_0    conda-forge
nbconvert                 5.5.0                      py_0    conda-forge
nbformat                  4.4.0                      py_1    conda-forge
ncurses                   6.1               h0a44026_1002    conda-forge
networkx                  2.3                        py_0    conda-forge
nltk                      3.2.5                      py_0    conda-forge
nomkl                     3.0                           0  
notebook                  5.7.8                    py36_1    conda-forge
numpy                     1.16.4           py36h6b0580a_0    conda-forge
oauthlib                  3.0.1                      py_0    conda-forge
olefile                   0.46                       py_0    conda-forge
openblas                  0.3.6                hd44dcd8_4    conda-forge
openssl                   1.1.1b               h01d97ff_2    conda-forge
palettable                3.1.1                      py_0    conda-forge
pandas                    0.24.2           py36h86efe34_0    conda-forge
pandoc                    2.7.3                         0    conda-forge
pandocfilters             1.4.2                      py_1    conda-forge
parso                     0.5.0                      py_0    conda-forge
pcre                      8.41              h0a44026_1003    conda-forge
pdfminer.six              20181108                   py_1    conda-forge
pexpect                   4.7.0                    py36_0    conda-forge
pickleshare               0.7.5                 py36_1000    conda-forge
pillow                    5.4.1           py36hbddbef0_1000    conda-forge
pip                       19.1.1                   py36_0    conda-forge
pkginfo                   1.5.0.1                    py_0    conda-forge
prometheus_client         0.7.1                      py_0    conda-forge
prompt_toolkit            2.0.9                      py_0    conda-forge
protobuf                  3.8.0            py36h6de7cb9_1    conda-forge
psutil                    5.6.3            py36h01d97ff_0    conda-forge
ptyprocess                0.6.0                   py_1001    conda-forge
py-lief                   0.9.0            py36h6d6d4d2_1    conda-forge
pycodestyle               2.5.0                      py_0    conda-forge
pycosat                   0.6.3           py36h1de35cc_1001    conda-forge
pycparser                 2.19                     py36_1    conda-forge
pycryptodome              3.8.2            py36h44b798e_0    conda-forge
pydispatcher              2.0.5                      py_1    conda-forge
pygments                  2.4.2                      py_0    conda-forge
pygpu                     0.7.6           py36h917ab60_1000    conda-forge
pyjwt                     1.7.1                      py_0    conda-forge
pymatgen                  2019.3.13                py36_0    matsci
pymongo                   3.8.0            py36h6de7cb9_0    conda-forge
pyopenssl                 19.0.0                   py36_0    conda-forge
pyparsing                 2.4.0                      py_0    conda-forge
pyqt                      5.9.2            py36h2a560b1_0    conda-forge
pyrsistent                0.15.3           py36h01d97ff_0    conda-forge
pysocks                   1.7.0                    py36_0    conda-forge
python                    3.6.7             h94afb7f_1005    conda-forge
python-crfsuite           0.9.6           py36h04f5b5a_1000    conda-forge
python-dateutil           2.8.0                      py_0    conda-forge
python-libarchive-c       2.8                   py36_1004    conda-forge
python-lmdb               0.94             py36hfc679d8_0    conda-forge
pytz                      2019.1                     py_0    conda-forge
pyyaml                    5.1.1            py36h01d97ff_0    conda-forge
pyzmq                     18.0.2           py36h2d07e9b_0    conda-forge
qt                        5.9.7                h93ee506_2    conda-forge
qtconsole                 4.5.1                      py_0    conda-forge
readline                  8.0                  hcfe32e1_0    conda-forge
regex                     2019.06.08       py36h01d97ff_0    conda-forge
requests                  2.22.0                   py36_0    conda-forge
requests-oauthlib         1.2.0                      py_0    conda-forge
ruamel                    1.0                      py36_0    conda-forge
ruamel.yaml               0.15.96          py36h01d97ff_0    conda-forge
ruamel_yaml               0.15.71         py36h1de35cc_1000    conda-forge
s3transfer                0.2.1                    py36_0    conda-forge
scikit-learn              0.21.2           py36hd4ffd6c_1    conda-forge
scipy                     1.3.0            py36hab3da7d_0    conda-forge
send2trash                1.5.0                      py_0    conda-forge
setuptools                41.0.1                   py36_0    conda-forge
sip                       4.19.8          py36h0a44026_1000    conda-forge
six                       1.12.0                py36_1000    conda-forge
smart_open                1.8.4                      py_0    conda-forge
sortedcontainers          2.1.0                      py_0    conda-forge
soupsieve                 1.9.2                    py36_0    conda-forge
spglib                    1.13.0           py36heacc8b8_0    conda-forge
sqlite                    3.28.0               hb7d70f7_1    conda-forge
sympy                     1.4                      py36_0    conda-forge
tabulate                  0.8.3                      py_0    conda-forge
tensorboard               1.13.1                   py36_0    conda-forge
tensorflow                1.13.1                   py36_0    conda-forge
tensorflow-estimator      1.13.0           py36h24bf2e0_0    conda-forge
termcolor                 1.1.0                      py_2    conda-forge
terminado                 0.8.2                    py36_0    conda-forge
testpath                  0.4.2                   py_1001    conda-forge
textblob                  0.15.3                     py_0    conda-forge
theano                    1.0.4           py36h0a44026_1000    conda-forge
tk                        8.6.9             h2573ce8_1002    conda-forge
toml                      0.10.0                     py_0    conda-forge
tornado                   6.0.3            py36h01d97ff_0    conda-forge
tqdm                      4.32.2                     py_0    conda-forge
traitlets                 4.3.2                 py36_1000    conda-forge
twython                   3.7.0                      py_0    conda-forge
unidecode                 1.1.1                      py_0    conda-forge
urllib3                   1.24.3                   py36_0    conda-forge
wcwidth                   0.1.7                      py_1    conda-forge
webencodings              0.5.1                      py_1    conda-forge
werkzeug                  0.15.4                     py_0    conda-forge
wheel                     0.33.4                   py36_0    conda-forge
widgetsnbextension        3.5.0                    py36_0    conda-forge
xz                        5.2.4             h1de35cc_1001    conda-forge
yaml                      0.1.7             h1de35cc_1001    conda-forge
zeromq                    4.3.1             h0a44026_1000    conda-forge
zlib                      1.2.11            h1de35cc_1004    conda-forge
zstd                      1.4.0                ha9f0a20_0    conda-forge
```

**Comment by lfoppiano** [comment]
I'm testing on linux now, 

I've got a missing library / failing tensorflow module. The process should crash. 

```
ERROR [2019-07-09 02:09:19,704] org.grobid.core.jni.JEPThreadPool: JEP initialization failed
! jep.JepException: <class 'ImportError'>: Traceback (most recent call last):
!   File "/home/Luca/.conda/envs/tensorflow/lib/python3.6/site-packages/tensorflow/python/pywrap_tensorflow.py", line 58, in <module>
!     from tensorflow.python.pywrap_tensorflow_internal import *
!   File "/home/Luca/.conda/envs/tensorflow/lib/python3.6/site-packages/tensorflow/python/pywrap_tensorflow_internal.py", line 28, in <module>
!     _pywrap_tensorflow_internal = swig_import_helper()
!   File "/home/Luca/.conda/envs/tensorflow/lib/python3.6/site-packages/tensorflow/python/pywrap_tensorflow_internal.py", line 24, in swig_import_helper
!     _mod = imp.load_module('_pywrap_tensorflow_internal', fp, pathname, description)
!   File "/home/Luca/.conda/envs/tensorflow/lib/python3.6/imp.py", line 243, in load_module
!     return load_dynamic(name, filename, file)
!   File "/home/Luca/.conda/envs/tensorflow/lib/python3.6/imp.py", line 343, in load_dynamic
!     return _load(spec)
! ImportError: /lib64/libstdc++.so.6: version `GLIBCXX_3.4.20' not found (required by /home/Luca/.conda/envs/tensorflow/lib/python3.6/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so)
! 
! 
! Failed to load the native TensorFlow runtime.
! 
! See https://www.tensorflow.org/install/errors
! 
! for some common reasons and solutions.  Include the entire stack trace
! above this error message when asking for help.
! at /home/Luca/.conda/envs/tensorflow/lib/python3.6/site-packages/tensorflow/python/pywrap_tensorflow.<module>(pywrap_tensorflow.py:74)
! at /home/Luca/.conda/envs/tensorflow/lib/python3.6/site-packages/tensorflow/python/__init__.<module>(__init__.py:49)
! at /home/Luca/.conda/envs/tensorflow/lib/python3.6/site-packages/tensorflow/__init__.<module>(__init__.py:24)
! at /home/Luca/.conda/envs/tensorflow/lib/python3.6/site-packages/keras/backend/tensorflow_backend.<module>(tensorflow_backend.py:5)
! at /home/Luca/.conda/envs/tensorflow/lib/python3.6/site-packages/keras/backend/__init__.<module>(__init__.py:89)
! at /home/Luca/.conda/envs/tensorflow/lib/python3.6/site-packages/keras/utils/conv_utils.<module>(conv_utils.py:9)
! at /home/Luca/.conda/envs/tensorflow/lib/python3.6/site-packages/keras/utils/__init__.<module>(__init__.py:6)
! at /home/Luca/.conda/envs/tensorflow/lib/python3.6/site-packages/keras/__init__.<module>(__init__.py:3)
! at <string>.<module>(<string>:1)
! at jep.Jep.eval(Native Method)
! at jep.Jep.eval(Jep.java:507)
! at org.grobid.core.jni.JEPThreadPool.initializeJepInstance(JEPThreadPool.java:86)
! at org.grobid.core.jni.JEPThreadPool.createJEPInstance(JEPThreadPool.java:106)
! at org.grobid.core.jni.JEPThreadPool.getJEPInstance(JEPThreadPool.java:136)
! at org.grobid.core.jni.DeLFTModel$InitModel.run(DeLFTModel.java:54)
! at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
! at java.util.concurrent.FutureTask.run(FutureTask.java:266)
! at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
! at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
! at java.lang.Thread.run(Thread.java:748)
Using TensorFlow backend.
name_header
running thread: 34
INFO  [2019-07-09 02:09:19,721] org.grobid.core.jni.DeLFTModel: Loading DeLFT model for name-header...
INFO  [2019-07-09 02:09:19,721] org.grobid.core.jni.JEPThreadPool: Creating JEP instance for thread 42
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007f4556448578, pid=6575, tid=0x00007f4556af0700
#
# JRE version: OpenJDK Runtime Environment (8.0_212-b04) (build 1.8.0_212-b04)
# Java VM: OpenJDK 64-Bit Server VM (25.212-b04 mixed mode linux-amd64 compressed oops)
# Problematic frame:
# C  [multiarray.cpython-36m-x86_64-linux-gnu.so+0x125578]  PyArray_Item_INCREF+0xe8
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /data/workspace/services/grobid-delft/hs_err_pid6575.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
```

**Comment by de-code** [comment]
I should say I generally only tested it with the non-CPU version of Tensorflow (as I don't have a GPU on my laptop). I did not experience the same crashes when activating a virtual env first. Is your JEP still installed within the virtual env? You could remove the one in `lin-64` just to be sure.

I am mostly using regular pip virtual envs. With conda env it was working for me as well. But when I tried to make it work in Docker it crashed silently (#441). I don't know what caused that.

The last error seem to be related to: https://stackoverflow.com/questions/41138179/tensorflow-build-fails-with-glibcxx-3-4-20-not-found

**Comment by lfoppiano** [comment]
@de-code the last issue of the record is that JEP initialisation is failing, however the service should exit by throwin up that exception, I think. 

That same error is due to the fact that Centos always results to be a pain when it's about glibc...

**Comment by de-code** [comment]
> @de-code the last issue of the record is that JEP initialisation is failing, however the service should exit by throwin up that exception, I think.

Yes, I agree, it would be better if it failed the startup.

> That same error is due to the fact that Centos always results to be a pain when it's about glibc...

Not sure if I can help there. That is why it's good to have Docker ;-)

**Comment by de-code** [comment]
How close are we to getting this merged? I suspect the tests are failing because the `grobid.properties` should configure `wapiti` instead of `delft`.

If it maybe will be a while, I am wondering whether we should extract some of the bug fixes out as a separate PR?

**Comment by lfoppiano** [comment]
@de-code the tests are failing because the GrobidProperties.getInstance() read what's in the `grobid.properties` and there is no way to force it toward one or another model. They should be independent from the configuration, else when you modify the configuration you cannot build it anymore without excluding the tests...

**Comment by kermitt2** [comment]
@lfoppiano You can set the "sequence labelling" machinery (WAPITI or DELFT) when creating a parser, for instance:

https://github.com/Impactstory/software-mentions/blob/master/src/main/java/org/grobid/core/engines/SoftwareParser.java#L85

```
private SoftwareParser() {
        super(GrobidModels.SOFTWARE, CntManagerFactory.getCntManager(),          GrobidCRFEngine.valueOf(SoftwareProperties.get("grobid.software.engine").toUpperCase()));
```

I think it's what's you are referring to? 

I've added this parameter when introducing DeLFT, so the test needs to be revisited!

**Comment by de-code** [comment]
> @de-code the tests are failing because the GrobidProperties.getInstance() read what's in the `grobid.properties` and there is no way to force it toward one or another model. They should be independent from the configuration, else when you modify the configuration you cannot build it anymore without excluding the tests...

Related to that, it would probably be good, if the local `grobid.properties` was excluded from git (to avoid accidental commits).

**Comment by lfoppiano** [comment]
Note to myself.. We should allow python 3.7 as well..

**Comment by lfoppiano** [comment]
Merged #460 here.

**Comment by lfoppiano** [comment]
I've just tested with Linux and the problem here is not longer a problem: https://github.com/kermitt2/grobid/pull/454#issuecomment-509459455 


I solved by 
> conda install libgcc

> export export LD_PRELOAD=/opt/anaconda3/lib/libstdc++.so.6.0.25

and it seems to work now.

File: doc/Deep-Learning-models.md [hunk 2]
```diff
@@ -4,28 +4,83 @@
...
+#### Classic python 
+
+- install [DeLFT](https://github.com/kermitt2/delft) 
 
 - indicate the path of the DeLFT install in `grobid.properties` (`grobid-home/config/grobid.properties`)
 
 - change the engine from `wapiti` to `delft` in the `grobid-properties` file
 
-Integration is realized via Java Embedded Python [JEP](https://github.com/ninia/jep), which uses a JNI of CPython. This integration is two times faster than the Tensorflow Java API and significantly faster than RPC serving (see https://www.slideshare.net/FlinkForward/flink-forward-berlin-2017-dongwon-kim-predictive-maintenance-with-apache-flink), and it does not require to modify DeLFT as it would be the case with Py4J gateway (socket-based).
+- run grobid
```
Comment (de-code) [inline doc/Deep-Learning-models.md]: I would add that the virtual environment should be activated before running GROBID?

Comment (lfoppiano) [reply doc/Deep-Learning-models.md]: we should add instruction for the virtualenv actually

**Pull request merged**